test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs

```diff
@@ -220,9 +220,102 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument document)
             var baseline = testFile.ReadAllText();
 
             // Normalize newlines to match those in the baseline.
-            var actual = serializedMappings.Replace("\r", "").Replace("\n", "\r\n");
+            var actualBaseline = serializedMappings.Replace("\r", "").Replace("\n", "\r\n");
 
-            Assert.Equal(baseline, actual);
+            Assert.Equal(baseline, actualBaseline);
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var visitor = new CodeSpanVisitor();
+            visitor.VisitBlock(syntaxTree.Root);
+
+            var charBuffer = new char[codeDocument.Source.Length];
+            codeDocument.Source.CopyTo(0, charBuffer, 0, codeDocument.Source.Length);
+            var sourceContent = new string(charBuffer);
+
+            var spans = visitor.CodeSpans;
+            for (var i= 0; i < spans.Count; i++)
+            {
+                var span = spans[i];
+                if (span.Start.FilePath == null || span.Start.FilePath != codeDocument.Source.FilePath)
+                {
+                    // Not in the main file, skip.
+                    continue;
+                }
+
+                var location = new SourceSpan(span.Start, span.Length);
+                var expectedSpan = sourceContent.Substring(span.Start.AbsoluteIndex, span.Length);
+
+                // See #
+                if (string.IsNullOrWhiteSpace(expectedSpan))
+                {
+                    // For now we don't verify whitespace inside of a directive. We know that directives cheat
+                    // with how they bound whitespace/C#/markup to make completion work.
+                    if (span.Parent is Block block && block.Type == BlockKindInternal.Directive)
+                    {
+                        continue;
+                    }
+                }
+
+                // See #
+                if (string.Equals("@", expectedSpan) && span.Kind == SpanKindInternal.Code)
+                {
+                    // For now we don't verify an escaped transition. In some cases one of the @ tokens in @@foo
+                    // will be mapped as C# but will not be present in the output buffer because it's not actually C#.
+                    continue;
+                }
+
+                var found = false;
+                for (var j = 0; j < csharpDocument.SourceMappings.Count; j++)
```

Comment:
Do we care if some source mappings aren't accounted for? Basically, right now you're checking each Span against all of the possible source mappings. It's possible we've generated an extra source mapping that incorrectly maps back to the SyntaxTree. Very possible with us manipulating the IR as much as we do.

Reply:
That's a good point. Let me see if it's easy to add.